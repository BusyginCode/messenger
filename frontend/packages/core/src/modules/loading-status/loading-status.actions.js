import { createAction } from 'redux-actions'

export const setEntityLoading = createAction('APP_SET_ENTITY_LOADING')

export const setEntityLoaded = createAction('APP_SET_ENTITY_LOADED')

export const setEntityFailed = createAction('APP_SET_ENTITY_FAILED')
