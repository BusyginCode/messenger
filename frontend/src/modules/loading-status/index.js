export * from './loading-status.actions'
export * from './loading-status.selectors'
export * from './loading-status.constants'

export { default } from './loading-status.reducer'
