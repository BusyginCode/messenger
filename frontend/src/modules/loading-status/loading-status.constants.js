export const loadingStatus = {
  NOT_STARTED: 'not-started',
  LOADING: 'loading',
  LOADED: 'loaded',
  FAILED: 'failed'
}
