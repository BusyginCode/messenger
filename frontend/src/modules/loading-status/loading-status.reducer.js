import { handleActions } from 'redux-actions'

import { setEntityLoading, setEntityLoaded, setEntityFailed } from './loading-status.actions'
import { loadingStatus } from './loading-status.constants'

const initialState = {}

const handleEntityLoading = (state, { payload: { entity, uid = 'list' } }) => ({
  ...state,
  [`${entity}/${uid}`]: loadingStatus.LOADING
})

const handleEntityLoaded = (state, { payload: { entity, uid = 'list' } }) => ({
  ...state,
  [`${entity}/${uid}`]: loadingStatus.LOADED
})

const handleEntityFailed = (state, { payload: { entity, uid = 'list' } }) => ({
  ...state,
  [`${entity}/${uid}`]: loadingStatus.FAILED
})

const reducer = handleActions(
  {
    [setEntityLoading]: handleEntityLoading,
    [setEntityLoaded]: handleEntityLoaded,
    [setEntityFailed]: handleEntityFailed
  },
  initialState
)

export default reducer
