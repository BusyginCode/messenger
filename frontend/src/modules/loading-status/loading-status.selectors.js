import get from 'lodash/get'
import includes from 'lodash/includes'
import upperFirst from 'lodash/upperFirst'
import { loadingStatus } from './loading-status.constants'

export const isEntityLoading = (state, entity, uid = 'list') => {
  return get(state, ['loadingStatus', `${entity}/${uid}`]) === loadingStatus.LOADING
}

export const isEntityLoaded = (state, entity, uid = 'list') => {
  return get(state, ['loadingStatus', `${entity}/${uid}`]) === loadingStatus.LOADED
}

export const isEntityFailed = (state, entity, uid = 'list') => {
  return get(state, ['loadingStatus', `${entity}/${uid}`]) === loadingStatus.FAILED
}

export const isEntityCompleted = (state, entity, uid = 'list') => {
  return includes([loadingStatus.LOADED, loadingStatus.FAILED], get(state, ['loadingStatus', `${entity}/${uid}`]))
}

export const isEntityNotStarted = (state, entity, uid = 'list') => {
  return includes([loadingStatus.INITIAL, undefined], get(state, ['loadingStatus', `${entity}/${uid}`]))
}

export const builder = (entity, options = {}) => {
  const { extended = true } = options
  return {
    [`is${extended ? upperFirst(entity) : ''}Loading`]: (state, uid = 'list') => isEntityLoading(state, entity, uid),
    [`is${extended ? upperFirst(entity) : ''}Loaded`]: (state, uid = 'list') => isEntityLoaded(state, entity, uid),
    [`is${extended ? upperFirst(entity) : ''}Failed`]: (state, uid = 'list') => isEntityFailed(state, entity, uid),
    [`is${extended ? upperFirst(entity) : ''}Completed`]: (state, uid = 'list') =>
      isEntityCompleted(state, entity, uid),
    [`is${extended ? upperFirst(entity) : ''}NotStarted`]: (state, uid = 'list') =>
      isEntityNotStarted(state, entity, uid)
  }
}
