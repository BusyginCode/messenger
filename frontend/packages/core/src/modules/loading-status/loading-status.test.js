import _each from 'lodash/map'
import { createMockStore } from '@messenger/testing'
import { setEntityLoading, setEntityLoaded, setEntityFailed } from './loading-status.actions'
import * as selectors from './loading-status.selectors'
import reducer from './loading-status.reducer'
import { loadingStatus } from './loading-status.constants'
const { isEntityNotStarted, isEntityLoading, isEntityLoaded, isEntityFailed, isEntityCompleted, builder } = selectors

describe('loadingStatus actions', () => {
  const store = createMockStore({ app: {}, auth: {} })

  describe('#setEntityLoading', () => {
    beforeEach(() => {
      store.clearActions()
    })

    it('should run action', () => {
      expect(store.dispatch(setEntityLoading({ entity: 'playersById', uid: '3j4kj34kjnjnkjn' }))).toMatchSnapshot()
    })
  })

  describe('#setEntityFailed', () => {
    beforeEach(() => {
      store.clearActions()
    })

    it('should run action', () => {
      expect(store.dispatch(setEntityFailed({ entity: 'playersById', uid: '3j4kj34kjnjnkjn' }))).toMatchSnapshot()
    })
  })

  describe('#setEntityLoaded', () => {
    beforeEach(() => {
      store.clearActions()
    })

    it('should run action', () => {
      expect(store.dispatch(setEntityLoaded({ entity: 'playersById', uid: '3j4kj34kjnjnkjn' }))).toMatchSnapshot()
    })
  })
})

describe('selectors', () => {
  const state = {
    loadingStatus: {
      'playersById/ent1': loadingStatus.INITIAL,
      'playersById/ent2': loadingStatus.LOADING,
      'playersById/ent3': loadingStatus.LOADED,
      'playersById/ent4': loadingStatus.FAILED
    }
  }

  describe('#isEntityLoading', () => {
    it('should return `true` in case entity has loading status', () => {
      _each([2], i => {
        expect(isEntityLoading(state, 'playersById', `ent${i}`)).toEqual(true)
      })
    })

    it('should return `false` in case entity has other status', () => {
      _each([0, 1, 3, 4], i => {
        expect(isEntityLoading(state, 'playersById', `ent${i}`)).toEqual(false)
      })
    })
  })

  describe('#isEntityLoaded', () => {
    it('should return `true` in case entity has loaded status', () => {
      _each([3], i => {
        expect(isEntityLoaded(state, 'playersById', `ent${i}`)).toEqual(true)
      })
    })

    it('should return `false` in case entity has other status', () => {
      _each([0, 1, 2, 4], i => {
        expect(isEntityLoaded(state, 'playersById', `ent${i}`)).toEqual(false)
      })
    })
  })

  describe('#isEntityFailed', () => {
    it('should return `true` in case entity has loaded status', () => {
      _each([4], i => {
        expect(isEntityFailed(state, 'playersById', `ent${i}`)).toEqual(true)
      })
    })

    it('should return `false` in case entity has other status', () => {
      _each([0, 1, 2, 3], i => {
        expect(isEntityFailed(state, 'playersById', `ent${i}`)).toEqual(false)
      })
    })
  })

  describe('#isEntityCompleted', () => {
    it('should return `true` in case entity has completed status', () => {
      _each([3, 4], i => {
        expect(isEntityCompleted(state, 'playersById', `ent${i}`)).toEqual(true)
      })
    })

    it('should return `false` in case entity has other status', () => {
      _each([0, 1, 2], i => {
        expect(isEntityCompleted(state, 'playersById', `ent${i}`)).toEqual(false)
      })
    })
  })

  describe('#isEntityNotStarted', () => {
    it('should return `true` in case entity has completed status', () => {
      _each([0, 1], i => {
        expect(isEntityNotStarted(state, 'playersById', `ent${i}`)).toEqual(true)
      })
    })

    it('should return `false` in case entity has other status', () => {
      _each([2, 3, 4], i => {
        expect(isEntityNotStarted(state, 'playersById', `ent${i}`)).toEqual(false)
      })
    })
  })

  describe('#build', () => {
    it('should return selector functions', () => {
      expect(builder('player')).toMatchSnapshot()
    })

    it('should return selector functions with short name in case extended: false option provided', () => {
      expect(builder('player', { extended: false })).toMatchSnapshot()
    })

    it('output functions should work exactly the same', () => {
      const builtFunctions = builder('playersById', { extended: false })
      const state = {
        loadingStatus: {
          'playersById/ent1': loadingStatus.INITIAL,
          'playersById/ent2': loadingStatus.LOADING,
          'playersById/ent3': loadingStatus.LOADED,
          'playersById/ent4': loadingStatus.FAILED
        }
      }
      _each(['Loading', 'Loaded', 'Failed', 'Completed', 'NotStarted'], stateId => {
        _each([0, 1, 2, 3, 4], i => {
          const builtFunctionResult = builtFunctions[`is${stateId}`](state, 'playersById', `ext${i}`)
          const functionResult = selectors[`isEntity${stateId}`](state, 'playersById', `ext${i}`)
          expect(builtFunctionResult).toEqual(functionResult)
        })
      })
    })

    it('output function with extended name should work the same as output function with short name', () => {
      const builtFunctionsExtended = builder('playersById')
      const builtFunctions = builder('playersById', { extended: false })
      const state = {
        loadingStatus: {
          'playersById/ent1': loadingStatus.INITIAL,
          'playersById/ent2': loadingStatus.LOADING,
          'playersById/ent3': loadingStatus.LOADED,
          'playersById/ent4': loadingStatus.FAILED
        }
      }
      _each(['Loading', 'Loaded', 'Failed', 'Completed', 'NotStarted'], stateId => {
        _each([0, 1, 2, 3, 4], i => {
          const extendedBuiltFunctionResult = builtFunctionsExtended[`isPlayersById${stateId}`](
            state,
            'playersById',
            `ext${i}`
          )
          const builtFunctionResult = builtFunctions[`is${stateId}`](state, 'playersById', `ext${i}`)
          expect(extendedBuiltFunctionResult).toEqual(builtFunctionResult)
        })
      })
    })
  })
})

describe('loadingStatus reducer', () => {
  it('should reduce setEntityLoading correctly', () => {
    const state = {}

    expect(
      reducer(state, {
        type: setEntityLoading,
        payload: { entity: 'player', uid: '343434' }
      })
    ).toMatchSnapshot()
  })
})

describe('loadingStatus reducer', () => {
  it('should reduce setEntityLoading correctly', () => {
    const state = {}

    expect(
      reducer(state, {
        type: setEntityLoaded,
        payload: { entity: 'player', uid: '343434' }
      })
    ).toMatchSnapshot()
  })
})

describe('loadingStatus reducer', () => {
  it('should reduce setEntityLoading correctly', () => {
    const state = {}

    expect(
      reducer(state, {
        type: setEntityFailed,
        payload: { entity: 'player', uid: '343434' }
      })
    ).toMatchSnapshot()
  })
})
