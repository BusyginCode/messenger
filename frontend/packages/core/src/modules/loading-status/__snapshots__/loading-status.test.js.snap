// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadingStatus actions #setEntityFailed should run action 1`] = `
Object {
  "payload": Object {
    "entity": "playersById",
    "uid": "3j4kj34kjnjnkjn",
  },
  "type": "APP_SET_ENTITY_FAILED",
}
`;

exports[`loadingStatus actions #setEntityLoaded should run action 1`] = `
Object {
  "payload": Object {
    "entity": "playersById",
    "uid": "3j4kj34kjnjnkjn",
  },
  "type": "APP_SET_ENTITY_LOADED",
}
`;

exports[`loadingStatus actions #setEntityLoading should run action 1`] = `
Object {
  "payload": Object {
    "entity": "playersById",
    "uid": "3j4kj34kjnjnkjn",
  },
  "type": "APP_SET_ENTITY_LOADING",
}
`;

exports[`loadingStatus reducer should reduce setEntityLoading correctly 1`] = `
Object {
  "player/343434": "loading",
}
`;

exports[`loadingStatus reducer should reduce setEntityLoading correctly 2`] = `
Object {
  "player/343434": "loaded",
}
`;

exports[`loadingStatus reducer should reduce setEntityLoading correctly 3`] = `
Object {
  "player/343434": "failed",
}
`;

exports[`selectors #build should return selector functions 1`] = `
Object {
  "isPlayerCompleted": [Function],
  "isPlayerFailed": [Function],
  "isPlayerLoaded": [Function],
  "isPlayerLoading": [Function],
  "isPlayerNotStarted": [Function],
}
`;

exports[`selectors #build should return selector functions with short name in case extended: false option provided 1`] = `
Object {
  "isCompleted": [Function],
  "isFailed": [Function],
  "isLoaded": [Function],
  "isLoading": [Function],
  "isNotStarted": [Function],
}
`;
